src/controlobjectscript.cpp

```diff
@@ -0,0 +1,55 @@
+#include <QApplication>
+#include <QtDebug>
+
+#include "controlobjectscript.h"
+
+ControlObjectScript::ControlObjectScript(const ConfigKey& key, QObject* pParent)
+        : ControlObjectSlave(key, pParent) {
+}
+
+void ControlObjectScript::connectScriptFunction(
+        const ControllerEngineConnection& conn) {
+    if (m_connectedScriptFunctions.isEmpty()) {
+        // we connect the slots only, if there will be actually a script
+        // connected
+        connect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)),
+                Qt::QueuedConnection);
+        connect(this, SIGNAL(trigger(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)),
+                Qt::QueuedConnection);
+    }
+    m_connectedScriptFunctions.append(conn);
+}
+
+bool ControlObjectScript::disconnectScriptFunction(
+        const ControllerEngineConnection& conn) {
+    bool ret = m_connectedScriptFunctions.removeAll(conn) > 0;
+    if (m_connectedScriptFunctions.isEmpty()) {
+        // no script left, we can disconnected
+        disconnect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)));
+        disconnect(this, SIGNAL(trigger(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)));
+    }
+    return ret;
+}
+
+void ControlObjectScript::slotValueChanged(double value, QObject*) {
+    // Make a local copy of m_connectedScriptFunctions fist.
+    // This allows a script to disconnect a callback from the callback
+    // itself. Otherwise the this may crash since the disconnect call
+    // happens during conn.function.call() in the middle of the loop below.
+    QList<ControllerEngineConnection> connections = m_connectedScriptFunctions;
+    for(auto&& conn: connections) {
```

Comment:
In this context auto&& is a universal reference, not an r-value reference:

https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

"Because the type deduction rules for |auto|-declared variables are 
essentially the same as for templates, it’s also possible to have 
|auto|-declared universal references."

On 01/13/2016 10:53 PM, Daniel Schürmann wrote:
>
> In src/controlobjectscript.cpp 
> <https://github.com/mixxxdj/mixxx/pull/853#discussion_r49656531>:
>
> > +        // no script left, we can disconnected
> > +        disconnect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
> > +                this, SLOT(slotValueChanged(double,QObject*)));
> > +        disconnect(this, SIGNAL(trigger(double, QObject*)),
> > +                this, SLOT(slotValueChanged(double,QObject*)));
> > +    }
> > +    return ret;
> > +}
> > +
> > +void ControlObjectScript::slotValueChanged(double value, QObject*) {
> > +    // Make a local copy of m_connectedScriptFunctions fist.
> > +    // This allows a script to disconnect a callback from the callback
> > +    // itself. Otherwise the this may crash since the disconnect call
> > +    // happens during conn.function.call() in the middle of the loop below.
> > +    QList<ControllerEngineConnection> connections = m_connectedScriptFunctions;
> > +    for(auto&& conn: connections) {
>
> I have just tested it and it works with just auto& as well.
> It seams to be that the auto&& is required for some std:: containers, but 
> not for QList.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/853/files#r49656531>.
>